### ISSUE TITLE
Implement Option 1: A `triggerBy` option that issues element re-render

### ISSUE THREAD

**Issue by jodator** [ckeditor/ckeditor5#7956]
## Provide a description of the task

Implement the base API for the [slot conversion API](https://github.com/ckeditor/ckeditor5/issues/6510#issuecomment-673417728). This should provide a minimal, top-level API for changes defined there and researched in: #7729.

### Goals:

*   top-level API as described, anything below should support it.
*   Use memoization technique to reduce re-rendering big chunks of the same views.
*   The renderer is out of scope of this task.

**Comment by jodator** [comment]
To have this properly handled from top to bottom I'll start with:

*   [ ] clearing up the code so the test will pass
*   [ ] write test cases for the feature - to define proper bahavior
*   [ ] update/refactor the code

**Comment by jodator** [comment]
I think that I have a solid base of test for the feature - as for now I don't think that any corner cases should be handled.

And while writing it down I see that I lack tests for not re-converting element for changes below slot, ie do not re-convert for changes in a `<paragraph>` inside a slot.

The other thing is that current API lacks some sort of the element binding (the case for outer/inner view elements maps to one model element - like all elements wrapped in `<figure>`). For now I'm thinking about extending `mapper.bind` to allow such mapping and, for now, will do exactly the same as what we do with double binding now.

**Comment by jodator** [comment]
I tried to make this work with table conversion. The base scenario works well and the code could be simplified a bit and more fluent in other parts. However, re-rendering on multiple changes (mixed with `triggerBy`) fails currently.

The current table refreshing logic is triggered for any `headingRows` attributes change, so in the new API it could be refactored to:

```javascript
conversion.for( 'editingDowncast' ).elementToElement( {
	model: 'table',
	view: downcastInsertTable( { asWidget: true } ),
	triggerBy: [
		'attribute:headingRows:table'					// <-- this should handle (1)
	]
} );
```

And `(1)\` is the whole: https://github.com/ckeditor/ckeditor5/blob/1cab8652830522f443dfd31570908a1c8ef9c52f/packages/ckeditor5-table/src/converters/table-heading-rows-refresh-post-fixer.js#L25 (the whole file is about 50 lines long).

This works nice in tests but also showed me that I didn't add more complex test cases. The problem in table scenarios is when:

1. refresh must be triggered ie by attribute change
2. **and** structure has changed also (added/removed child) _and_ is not defined in `triggerBy` (might not be relevant).

Why? I didn't know differ internals - there's a huuuuge if-else tree (I'd refactor this to be purely funcitonal maybe some day) that handles overlapping changes. 

For instance, if there's an `insert` change that overlaps an existing `remove` change,  the `remove` change is deleted from changes set (but not going deep into what actually happens there). Now, for proposed `refresh` change we need to:

1. have a similar way of deleting changes that will be handled by the converter anyway
2. while allowing parts of the tree _below_ the changed element (slots content).

**Comment by jodator** [comment]
For starters, I'll try to write tests for similar cases in the engine and see if those shed some light on the issue or solutions. We might invalidate this case for now and revert table changes.

The other thing is to implement `triggerBy` callbacks together with a complete set of events and a way to manipulate them.

Alternatively, (to not make the solution too complex to use from outside) information about "slot" should be available somewhere (it is in the mapper for instance) so I'll check that first. The goal is to not bloat the API from the inside and to figure things out internally.

**Comment by jodator** [comment]
> For starters, I'll try to write tests for similar cases in the engine and see if those shed some light on the issue or solutions. We might invalidate this case for now and revert table changes.

Oh, I have a similar failing test for this already.

**Comment by jodator** [comment]
I was trying to handle complex cases of refresh events, a scenario that was kind of hard to figure out was:

```
$root
	complex		<- refresh, ie by attribute change
		slot	
		slot	<- remove
```

with configuration like:

```javascript
conversion.for( 'downcast' ).elementToElement( {
	model: 'complex',
	view: () => { ... },
	triggerBy: [
		'attribute:foo:complex',
		'insert:slot',
		'remove:slot'
	]
} );
```

What should happen is that `remove` operation is not handled by the dispatcher, or in other words a set of changes from a differ should only include `refresh` change for the `<complex>` only.

Handling this on a `Differ` level turned out to be pointless as Differ can't be aware of when given child change should be hindered by the `refresh` change and when not (for instance in the above example change below `<slot>` (for instance `<paragraph>` inserted in `<slot>`) should not be hidden by `refresh` of the `complex` element and should be converted.

However, the downcast dispatcher has this information as it handles `triggerBy` options.

My next steps will be to skip conversion of matched events either by:

1.  marking some changes as outdated in the differ (so the differ would return only `refresh`)
2.  skipping differ results for matched events

As for now the 1) seems a bit to much attached to differ while 2) looks like a proper place to handle above.

**Comment by jodator** [comment]
I've manage to dig out some issue that still need to be resolved, most likely covering the above post.

*   consuming in the general `insert` converter on "refresh" was too greedy
*   proper events handling to check what needs to converter is needed in refresh handling
    *   do not convert items outside refreshed element (figure it out on events)
    *   convert new items outside refreshed element (as above)

**Comment by jodator** [comment]
The full reconversion with slots is kinda working. The case to handle is a hybrid approach in which main element is handled by reconversion while adding/removing slots is done separately. If time allow I'll investigate this, however I'm worried that it would require API for a layered view items mapping (`image -> figure>img` case).

**Comment by jodator** [comment]
So for so good. I've "fixed" the API for now, added some docs and removed debug code. Still to handle:

*   [ ] table conversion - tests are failing after `triggerBy` was use in table converter (might be blocked by #1589)
*   [ ] engine code coverage dropped - inspect whether test are needed or there is dead code
*   [ ] internal API - code is a mess still in prototyped code

**Comment by jodator** [comment]
After some battling with table tests, it turned out that I've implemented the POC on the downcast helpers level. Which is not ideal. Table tests showed that removing old view and view momoization should be handled on the dispatcher level. The current form creates duplicated views when refreshing paragraphs in table cell.

**Comment by jodator** [comment]
Most of the issues with table conversion were related to the refreshing table cell. I've refactored that part to refresh `<paragraph>` in those cells only. This brought some changes to conversion (mostly due to special handling of the single paragraph in editing and data views) but the whole re-converting algorithm works :ok\_hand: here:

*   you can refresh the table only on specific events.
*   you can refresh paragraph in a refreshed table cell (I hope we had a test for that :D ).
*   you can have attribute converters on a refreshed element.

**Comment by jodator** [comment]
The #1589 is a blocker for a group of tests in the engine. So without it, you could only write a whole-refresh item converter that would include creating slots.

Tables have custom row insertion logic that does not rely on position mapping.

**Comment by jodator** [comment]
ATM the engine tests are OK with one exception for failing table tests.

In tables I have 3 failing tests and 5 ignored.

Failing tests requires some investigation of the downcast conversion ,- minor issue.

However ignored tests are worse as those cause browser to freeze as differ produces infinite loop condition in `getChanges()`. There's `unshift` used in the loop processing an array and in those cases, it unshifts, the same change indifinietly.

All of those changes might be tied together. I'd suspect some glitch in table conversion (maybe refresh() is used too often) or the differ still needs and improvement in order to properly handle this edge case.

I'll try to investigate what set of changes produces the infinite loop.

**Comment by jodator** [comment]
_\# Re-converting whole thing + slot conversion_  
  
Table feature showed me shortcomings of what I though was needed.  
  
_\## Cases:_  
  
1\. Element without children (w, w/o external \`attributeToAttribute()\` conversion)  
  
Always replaced by a new view structure. Relatively simple.  
  
\`\`\`html  
\<apple type="macintosh">\</apple>  
  
// converted to:  
  
\<div class="apple">\<span class="apple-type">macintosh\</span>\</div>  
\`\`\`  
  
Current use case: none.  
  
2\. Flat element with children (w, w/o external \`attributeToAttribute()\` conversion).  
  
\`\`\`html  
\<tree hasBugs="true">                           // Element start, children are here  
    \<paragraph>A\</paragraph>  
    \<paragraph alignment="right">B\</paragraph>  
\</tree>                                         // Element end  
  
// converted to:  
  
\<div class="tree tree\_with-bugs">               // Element start, content goes here  
    \<p>A\</p>  
    \<p style="text-align: right">A\</p>  
\</div>                                          // Element start  
\`\`\`  
  
Current use case: \`\<paragraph>\` inside tableCell requires specific rendering (either \`\<p>\` or \`\<span>\` in the editing view).  
  
3\. Element with children, complex view (no slots)  
  
\`\`\`html  
\<tower levels="2" type="pretty">                  // Element start, children are here  
    \<paragraph>A\</paragraph>  
    \<paragraph alignment="right">B\</paragraph>  
\</tower>                                         // Element end  
  
// converted to:  
  
\<div class="tower tower-pretty">               // Element start  
    \<div class="tower-level-1">  
        \<div class="tower-level-2">            // Content goes here  
            \<p>A\</p>  
            \<p style="text-align: right">A\</p>  
        \</div>  
    \</div>  
\</div>                                          // Element start  
\`\`\`  
  
Current use case: The \`\<image>\` to some extent.  
  
4\. Element with virtual "slots"  
  
Here slots not wrapped in a particular wrapper element or other elements might be rendered alongside them.  
  
\`\`\`html  
\<train type="steam">                        // Element start, slots are here  
    \<car>  
        \<paragraph>A\</paragraph>  
    \</car>  
    \<car>  
        \<paragraph alignment="right">B\</paragraph>  
    \</car>  
\</train>                                    // Element end  
  
// converted to:  
  
\<div class="train train-steam">             // Element start  
    \<div class="tender">\<span class="coal">\</span>\</div>  
    \<div class="car">  
        \<p>A\</p>  
    \</div>  
    \<div class="car">  
        \<p style="text-align: right">A\</p>  
    \</div>  
\</div>                                      // Element end  
\`\`\`  
  
Demo with \`\<tile>\` and \`\<body>\` sections. The \`\<table>\` to some extent. Also \`\<table>\` and \`\<mediaEmbed>\` if you with to add \`\<caption>\` to them as an external converter.  
  
The problem here is that each slot must be fully controlled by the main converter. You can not add an \`attributeToAttribue()\` converter to the slot in this case (at least easily).  
  
The tables turned to be a tricky - slots at \`\<tableCell>\` level are extremly hard as other, granular table features needs to convert its attributes. Thus \`\<slot>\` in terms of tables can be:  
\- on \`\<tableRow>\` level - feasible in 2-4 days. table converter insert each row (engine tests are already covering this case)  
\- on \`table -> thead,tbody\` level - I'm not sure if feasible in short period of time. Specific use-case in which \*one\* model element has \*two+\* view elements. Probably would require model-to-view position mapping. Interesting idea but requires low-level aproach and I don't see an easy API for that. The slot conversion of table+tableRow as in previous point is good as we will not have to translate some positions.  
  
_\### Table issues_  
  
Hard to draw a valid line in tables. At first it seemed that \`\<tableCell>\` is a slot. It is not. More on this below.  
  
In order to implement \`triggerBy\` I've narrowed down the scope to single, flat item re-conversion and used it in table cell paragraph refreshing (current table cell refresh post-fixer). It refreshes only the paragraphs inside table cell as refreshing whole table cell seems redundant.  
  
Smart refreshing of a whole table cell is possible but requires a table converter refactor in order to properly mark what changed and what is needed to re-render.  
  
There is a possibility to have \`\<tableCell>\` as "slot as we can fire \`attribute\` converters even on already rendered element.

**Comment by jodator** [comment]
Case 1: The less problematic case IMO - self contained element without children - tested, works.

Case 2: I implemented in #8099 as `differ.refreshItem( paragraph )` in table cell refresh post-fixer (previously you would refresh the whole table cell - I don't see that necessary).

Case 3: Requires #1589 - an API to tell the mapper where elments start/ends in the view tree. Partially handled by double-bindings in some places but it is buggy.

Case 4: I'd postpone this case for own ticket. Havin #1589 should easy implementing this. Additionally attributes conversion for rendered `<slot>` should be also possible (table properties scenario).

**Comment by jodator** [comment]
To conclude this ticekt with a minimal use case I'd go with enabling the most obvious solutions for reconversion - mostly as a demo/feature guides. Other things should be fixed by the follow-ups to unblock refreshing a table with minimal view changes and allowing to implement `<caption>` insertion to `<mediaEmbed>` like describe in #2780.

**Comment by jodator** [comment]
So the remaining thing is to fix memoization used together with `triggerBy` (and future element reconverting API - a.k.a. new `Differ.refreshItem()`. 

It turned out that I've implemented view memoization only when the "slots" are defined and missed the simplest case where a block element is refreshed and it's children should be retained.

The `differ.rereshItem()` changes were reverted as it turned out that the whole reconversion can be (should be?) handled by the downcast disptacher (starting from this [PR comment](https://github.com/ckeditor/ckeditor5/pull/8099#issuecomment-705414990)).

Changes to the `table-cell-refresh-post-fixer` turned out te be OK - despite reverting to the old differ logic. Now only `<p>` is re-inserted rather than whole `<td>`.

ATM I'm going to fix the memoization logic to make those tests green ( 7e359d8b2967de4b0afd375254f038744eca5d52 ).

**Comment by jodator** [comment]
So the memoization works correctly in the downcast dispatcher and the logic should be simpler as the last thing to be removed was slot binding from the mapper.

The code changes are now restricted to `DowncastDispatcher` and `DowncastHelpers`.
